plugin/builtin/patchtransformer/PatchTransformer.go

```diff
@@ -143,9 +143,15 @@ func (p *plugin) applySMPatch(resource, patch *resource.Resource) error {
 	if err != nil {
 		return err
 	}
-	return filtersutil.ApplyToJSON(patchstrategicmerge.Filter{
+	n, ns := resource.GetName(), resource.GetNamespace()
+	err = filtersutil.ApplyToJSON(patchstrategicmerge.Filter{
 		Patch: node,
 	}, resource)
+	if resource.GetName() != "" {
```

Comment:
@monopole Do we have a method in `Resource` to check is the resource empty?

Reply:
I don't find it in the code but maybe I miss something.

Reply:
I also was looking for something like that but didn't find it which is why I did this instead. I may also have missed something.

Reply:
We don't have that concept.

namespace is optional, but kind, apiVersion and metadata/name are not.  An empty name makes the `kustomize build`, and moreover, the apiserver demands names too.

I think all we need here are the writes.  Restore the old values without even looking at them.
i.e. no `if` statement

Reply:
I'm tempted to blindly restore the value of the `kind` and `apiVersion` fields as well, but we don't have a specific mutator for them like we do for the `metadata/name` field.

Reply:
Removing the if statement causes the TestPatchTransformerWithPatchDelete test to fail, which is why I added it. What do you want to do for that test case?

Reply:
Sigh.  Yes, the deletion hack.  So we need the test, and now i better understand what @Shell32-Natsu was hinting at - we can have an invalid (deleted but zombie-like) resource come in here, and we shouldn't mess with it.

Reply:
`L125 if len(res.Map()) == 0 {`
is the closest thing we have to an "IsEmpty" test.
So we could test that here, which would be clearer thant just looking at the name field.
i'm also curious if it would work.
then in a later PR someone could introduce IsEmpty on the Resource interface, and use it everywhere where we test the length of the Map().

Reply:
updated